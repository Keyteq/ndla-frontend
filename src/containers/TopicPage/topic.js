/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { handleActions, createAction } from 'redux-actions';
import { createSelector } from 'reselect';
import defined from 'defined';

import groupBy from '../../util/groupBy';
import { getArticle } from '../ArticlePage/article';
import { getArticleIdFromResource } from '../Resources/resourceHelpers';
import createFetchActions from '../../util/createFetchActions';

export const fetchTopicsActions = createFetchActions('TOPICS');
export const fetchTopicsWithIntroductions = createAction(
  'FETCH_TOPICS_WITH_INTRODUCTIONS',
);
export const setTopics = createAction('SET_TOPICS');
export const fetchTopicArticleActions = createFetchActions('TOPIC_ARTICLE');

export const setTopicIntroductions = createAction('SET_TOPIC_INTRODUCTIONS');

export const actions = {
  ...fetchTopicsActions,
  ...fetchTopicArticleActions,
  fetchTopicsWithIntroductions,
  setTopics,
  setTopicIntroductions,
};

export const initialState = {
  all: {},
  fetchTopicsStatus: 'initial',
  fetchTopicArticleStatus: 'initial',
  topicIntroductions: {},
};

export default handleActions(
  {
    [setTopics]: {
      next: (state, action) => {
        const { subjectId, topics } = action.payload;
        return {
          ...state,
          fetchTopicsStatus: 'success',
          all: { ...state.all, [subjectId]: topics },
        };
      },
      throw: state => state,
    },
    [setTopicIntroductions]: {
      next: (state, action) => {
        const { articleIntroductions, topics } = action.payload;
        // Map article introduction to topic
        const topicIntroductions = topics.reduce((obj, item) => {
          const article = articleIntroductions.find(
            articleIntroduction =>
              item.contentUri === `urn:article:${articleIntroduction.id}`,
          );
          return article ? { ...obj, [item.id]: article.introduction } : obj;
        }, {});

        return {
          ...state,
          topicIntroductions: {
            ...state.topicIntroductions,
            ...topicIntroductions,
          },
        };
      },
      throw: state => state,
    },
    [actions.fetchTopicsError]: {
      next: state => ({
        ...state,
        fetchTopicsStatus: 'error',
      }),
      throw: state => state,
    },
    [actions.fetchTopicArticle]: {
      next: state => ({
        ...state,
        fetchTopicArticleStatus: 'loading',
      }),
      throw: state => state,
    },
    [actions.fetchTopicArticleError]: {
      next: state => ({
        ...state,
        fetchTopicArticleStatus: 'error',
      }),
      throw: state => state,
    },
    [actions.fetchTopicArticleSuccess]: {
      next: state => ({
        ...state,
        fetchTopicArticleStatus: 'success',
      }),
      throw: state => state,
    },
  },
  initialState,
);

const getTopicsFromState = state => state.topics;

export const getTopicIntroductions = createSelector(
  [getTopicsFromState],
  topics => topics.topicIntroductions,
);

export const hasFetchedTopicsBySubjectId = subjectId =>
  createSelector(
    [getTopicsFromState],
    topics => topics.all[subjectId] !== undefined,
  );

export const getFetchTopicsStatus = createSelector(
  [getTopicsFromState],
  topics => topics.fetchTopicsStatus,
);

export const getFetchTopicArticleStatus = createSelector(
  [getTopicsFromState],
  topics => topics.fetchTopicArticleStatus,
);

export const getAllTopicsBySubjectId = subjectId =>
  createSelector([getTopicsFromState], topics =>
    defined(topics.all[subjectId], []),
  );

export const getTopicsBySubjectId = subjectId =>
  createSelector([getAllTopicsBySubjectId(subjectId)], topics =>
    topics.filter(topic => !topic.parent || topic.parent === subjectId),
  );

export const getTopicsBySubjectIdWithIntroduction = subjectId =>
  createSelector(
    [getTopicsBySubjectId(subjectId), getTopicIntroductions],
    (topics, topicIntroductions) =>
      topics.map(topic => {
        if (topic && topicIntroductions) {
          const topicIntroduction = topicIntroductions[topic.id];
          const introduction = topicIntroduction
            ? topicIntroduction.introduction
            : undefined;
          return { ...topic, introduction };
        }
        return topic;
      }),
  );

export const getTopic = (subjectId, topicId = undefined) =>
  createSelector([getAllTopicsBySubjectId(subjectId)], topics =>
    topics.find(topic => topicId === topic.id),
  );

export const getTopicArticle = (subjectId, topicId) =>
  createSelector(
    [getTopic(subjectId, topicId), state => state],
    (topic, state) =>
      topic && topic.contentUri
        ? getArticle(getArticleIdFromResource(topic))(state)
        : undefined,
  );

export const getSubtopics = (subjectId, topicId) =>
  createSelector([getAllTopicsBySubjectId(subjectId)], topics =>
    topics.filter(topic => topicId === topic.parent),
  );

export const getSubjectMenu = subjectId =>
  createSelector([getAllTopicsBySubjectId(subjectId)], topics => {
    const groupedSubtopicsByParent = groupBy(
      topics.filter(topic => topic.parent),
      'parent',
    );

    const toMenu = topic => {
      const subtopics = defined(groupedSubtopicsByParent[topic.id], []);

      const subtopicsWithSubtopics = subtopics.map(child => toMenu(child));

      return { ...topic, subtopics: subtopicsWithSubtopics };
    };

    return topics
      .filter(t => !t.parent || t.parent === subjectId)
      .map(root => toMenu(root));
  });

export const getTopicPath = (subjectId, topicId) =>
  createSelector(
    [getTopic(subjectId, topicId), getAllTopicsBySubjectId(subjectId)],
    (leaf, topics) => {
      if (!leaf) {
        return [];
      }

      const toBreadcrumb = topic => {
        if (!topic.parent || topic.parent === subjectId) {
          return [topic];
        }
        const parent = topics.find(t => topic.parent === t.id);
        const parentPath = toBreadcrumb(parent);
        return [...parentPath, topic];
      };

      const topicPath = toBreadcrumb(leaf);

      return topicPath; // Remove last item (leaf topic)
    },
  );

export const getSubtopicsWithIntroduction = (subjectId, topicId) =>
  createSelector(
    [getSubtopics(subjectId, topicId), getTopicIntroductions],
    (topics, topicIntroductions) =>
      topics.map(topic => {
        if (topic && topicIntroductions) {
          const topicIntroduction = topicIntroductions[topic.id];
          const introduction = topicIntroduction
            ? topicIntroduction.introduction
            : undefined;
          return { ...topic, introduction };
        }
        return topic;
      }),
  );
