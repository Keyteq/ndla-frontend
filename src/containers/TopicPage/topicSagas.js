/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { take, call, put, select } from 'redux-saga/effects';
import {
  actions,
  getTopic,
  hasFetchedTopicsBySubjectId,
  getAllTopicsBySubjectId,
} from './topic';
import { getLocale } from '../Locale/localeSelectors';
import {
  getArticleIdFromResource,
  isArticleResource,
} from '../Resources/resourceHelpers';
import { actions as articleActions, getArticle } from '../ArticlePage/article';
import { applicationError } from '../../modules/error';
import * as articleApi from '../ArticlePage/articleApi';
import * as api from './topicApi';

export function* fetchTopicIntroductions(topics) {
  try {
    const ids = topics.filter(isArticleResource).map(getArticleIdFromResource);

    if (ids.length === 0) {
      return;
    }
    const data = yield call(articleApi.fetchArticles, ids);
    yield put(
      actions.setTopicIntroductions({
        topics,
        articleIntroductions: data.results,
      }),
    );
  } catch (error) {
    yield put(applicationError(error));
  }
}

export function* fetchTopicArticle(subjectId, topicId) {
  try {
    const locale = yield select(getLocale);
    let topic = yield select(getTopic(subjectId, topicId));
    if (!topic) {
      topic = yield call(api.fetchTopic, topicId, locale);
    }
    const articleId = getArticleIdFromResource(topic);
    let article = yield select(getArticle(articleId));
    if (articleId && !article) {
      article = yield call(articleApi.fetchArticle, articleId, locale);
      yield put(articleActions.setArticle(article));
    }
    yield put(actions.fetchTopicArticleSuccess());
  } catch (error) {
    yield put(applicationError(error));
    yield put(actions.fetchTopicArticleError());
  }
}

export function* fetchTopics(subjectId) {
  try {
    const locale = yield select(getLocale);
    const topics = yield call(api.fetchTopics, subjectId, locale);
    yield put(actions.setTopics({ topics, subjectId }));
    return topics;
  } catch (error) {
    yield put(applicationError(error));
    yield put(actions.fetchTopicsError());
    return [];
  }
}

export function* fetchTopicsWithIntroductions(subjectId) {
  const hasFetched = yield select(hasFetchedTopicsBySubjectId(subjectId));
  if (!hasFetched) {
    const topics = yield call(fetchTopics, subjectId);
    yield call(fetchTopicIntroductions, topics);
  } else {
    const topics = yield select(getAllTopicsBySubjectId(subjectId));
    yield call(fetchTopicIntroductions, topics);
  }
}

export function* watchFetchTopicsWithIntroductions() {
  while (true) {
    const { payload: { subjectId } } = yield take(
      actions.fetchTopicsWithIntroductions,
    );
    yield call(fetchTopicsWithIntroductions, subjectId);
  }
}

export function* watchFetchTopics() {
  while (true) {
    const { payload: { subjectId } } = yield take(actions.fetchTopics);
    const hasFetched = yield select(hasFetchedTopicsBySubjectId(subjectId));
    if (!hasFetched) {
      yield call(fetchTopics, subjectId);
    }
  }
}

export function* watchFetchTopicArticle() {
  while (true) {
    const { payload: { subjectId, topicId } } = yield take(
      actions.fetchTopicArticle,
    );
    yield call(fetchTopicArticle, subjectId, topicId);
  }
}

export default [
  watchFetchTopics,
  watchFetchTopicsWithIntroductions,
  watchFetchTopicArticle,
];
