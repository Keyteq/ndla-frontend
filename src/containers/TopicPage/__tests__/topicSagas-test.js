/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { testSaga, expectSaga } from 'redux-saga-test-plan';
import nock from 'nock';
import * as sagas from '../topicSagas';
import { initialState, actions } from '../topic';
import { topics, articles } from './mockTopics';

test('resourceSagas watchFetchTopicsWithIntroductions', () => {
  const req1 = nock('http://ndla-api')
    .get(
      '/taxonomy/v1/subjects/urn:subject:1/topics/?recursive=true&language=en',
    )
    .reply(200, topics);

  const req2 = nock('http://ndla-api')
    .get('/article-api/v2/articles?ids=1,1_2')
    .reply(200, { results: articles });

  return expectSaga(sagas.watchFetchTopicsWithIntroductions)
    .withState({ topics: initialState, locale: 'en' })
    .dispatch(
      actions.fetchTopicsWithIntroductions({
        subjectId: 'urn:subject:1',
        topicId: 'urn:topic:1',
      }),
    )
    .silentRun(500)
    .then(result => {
      expect(result.toJSON()).toMatchSnapshot();
      req1.done();
      req2.done();
    });
});

test('topicSagas watchFetchTopicsWithIntroductions when topics are in state', () => {
  const req1 = nock('http://ndla-api')
    .get('/article-api/v2/articles?ids=1,1_2')
    .reply(200, { results: articles });

  return expectSaga(sagas.watchFetchTopicsWithIntroductions)
    .withState({ topics: { all: { 'urn:subject:1': topics } }, locale: 'en' })
    .dispatch(
      actions.fetchTopicsWithIntroductions({
        subjectId: 'urn:subject:1',
        topicId: 'urn:topic:1',
      }),
    )
    .silentRun(500)
    .then(result => {
      expect(result.toJSON()).toMatchSnapshot();
      req1.done();
    });
});

test('topicSagas watchFetchTopics', () => {
  const req1 = nock('http://ndla-api')
    .get(
      '/taxonomy/v1/subjects/urn:subject:1/topics/?recursive=true&language=en',
    )
    .reply(200, topics);

  return expectSaga(sagas.watchFetchTopics)
    .withState({ topics: initialState, locale: 'en' })
    .dispatch(
      actions.fetchTopics({
        subjectId: 'urn:subject:1',
      }),
    )
    .silentRun(500)
    .then(result => {
      expect(result.toJSON()).toMatchSnapshot();
      req1.done();
    });
});

test('topicSagas watchFetchTopics when topics already are in state', () =>
  expectSaga(sagas.watchFetchTopics)
    .withState({ topics: { all: { 'urn:subject:1': topics } } })
    .dispatch(
      actions.fetchTopics({
        subjectId: 'urn:subject:1',
      }),
    )
    .silentRun(500)
    .then(result => {
      expect(result.toJSON()).toMatchSnapshot();
    }));

test('topicSagas fetchTopicIntroductions do not call fetchArticles if no valid ids', () => {
  const mockTopics = [{ contentUri: 'urn:learningpath:2' }];
  const saga = testSaga(sagas.fetchTopicIntroductions, mockTopics);
  saga.next().isDone();
});

test('resourceSagas watchFetchTopicArticle', () => {
  const req1 = nock('http://ndla-api')
    .get('/taxonomy/v1/topics/urn:topic:1/?language=en')
    .reply(200, topics[0]);

  const req2 = nock('http://ndla-api')
    .get('/article-converter/json/en/1')
    .reply(200, {});

  return expectSaga(sagas.watchFetchTopicArticle)
    .withState({ topics: initialState, articles: { all: {} }, locale: 'en' })
    .dispatch(
      actions.fetchTopicArticle({
        subjectId: 'urn:subject:1',
        topicId: 'urn:topic:1',
      }),
    )
    .silentRun(500)
    .then(result => {
      expect(result.toJSON()).toMatchSnapshot();
      req1.done();
      req2.done();
    });
});
