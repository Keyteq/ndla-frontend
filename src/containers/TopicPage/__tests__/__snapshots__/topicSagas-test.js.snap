// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resourceSagas watchFetchTopicArticle 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "urn:subject:1",
          "urn:topic:1",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/fetchTopicArticle",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {},
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "urn:topic:1",
          "en",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "1",
          "en",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": Object {},
          "type": "SET_ARTICLE",
        },
        "channel": null,
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "type": "FETCH_TOPIC_ARTICLE_SUCCESS",
        },
        "channel": null,
      },
    },
  ],
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "SELECT": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "SELECT": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "SELECT": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "@@redux-saga-test-plan/json/function/actionCreator",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "@@redux-saga-test-plan/json/function/actionCreator",
      },
    },
  ],
}
`;

exports[`resourceSagas watchFetchTopicsWithIntroductions 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "urn:subject:1",
          undefined,
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/fetchTopicsWithIntroductions",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {},
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "urn:subject:1",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/fetchTopics",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {},
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "urn:subject:1",
          "en",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Array [
            Object {
              "contentUri": "urn:article:1",
              "id": "urn:topic:1",
              "name": "Idéutvikling og mediedesign",
              "parent": "urn:subject:1",
            },
            Object {
              "id": "urn:topic:1_1",
              "name": "Mediedesign",
              "parent": "urn:topic:1",
            },
            Object {
              "contentUri": "urn:article:1_2",
              "id": "urn:topic:1_2",
              "name": "Idéutvikling",
              "parent": "urn:topic:1",
            },
            Object {
              "id": "urn:topic:1_2_1",
              "name": "Mediebransjen",
              "parent": "urn:topic:1_2",
            },
            Object {
              "id": "urn:topic:2",
              "name": "Mediekommunikasjon",
            },
          ],
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/fetchTopicIntroductions",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {},
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Array [
            "1",
            "1_2",
          ],
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": Object {
            "subjectId": "urn:subject:1",
            "topics": Array [
              Object {
                "contentUri": "urn:article:1",
                "id": "urn:topic:1",
                "name": "Idéutvikling og mediedesign",
                "parent": "urn:subject:1",
              },
              Object {
                "id": "urn:topic:1_1",
                "name": "Mediedesign",
                "parent": "urn:topic:1",
              },
              Object {
                "contentUri": "urn:article:1_2",
                "id": "urn:topic:1_2",
                "name": "Idéutvikling",
                "parent": "urn:topic:1",
              },
              Object {
                "id": "urn:topic:1_2_1",
                "name": "Mediebransjen",
                "parent": "urn:topic:1_2",
              },
              Object {
                "id": "urn:topic:2",
                "name": "Mediekommunikasjon",
              },
            ],
          },
          "type": "SET_TOPICS",
        },
        "channel": null,
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": Object {
            "articleIntroductions": Array [
              Object {
                "id": "urn:article:1",
                "introduction": "Introduction 1",
              },
              Object {
                "id": "urn:article:1_2",
                "introduction": "Introduction 2",
              },
            ],
            "topics": Array [
              Object {
                "contentUri": "urn:article:1",
                "id": "urn:topic:1",
                "name": "Idéutvikling og mediedesign",
                "parent": "urn:subject:1",
              },
              Object {
                "id": "urn:topic:1_1",
                "name": "Mediedesign",
                "parent": "urn:topic:1",
              },
              Object {
                "contentUri": "urn:article:1_2",
                "id": "urn:topic:1_2",
                "name": "Idéutvikling",
                "parent": "urn:topic:1",
              },
              Object {
                "id": "urn:topic:1_2_1",
                "name": "Mediebransjen",
                "parent": "urn:topic:1_2",
              },
              Object {
                "id": "urn:topic:2",
                "name": "Mediekommunikasjon",
              },
            ],
          },
          "type": "SET_TOPIC_INTRODUCTIONS",
        },
        "channel": null,
      },
    },
  ],
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "SELECT": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "SELECT": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "@@redux-saga-test-plan/json/function/actionCreator",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "@@redux-saga-test-plan/json/function/actionCreator",
      },
    },
  ],
}
`;

exports[`topicSagas watchFetchTopics 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "urn:subject:1",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/fetchTopics",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {},
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "urn:subject:1",
          "en",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": Object {
            "subjectId": "urn:subject:1",
            "topics": Array [
              Object {
                "contentUri": "urn:article:1",
                "id": "urn:topic:1",
                "name": "Idéutvikling og mediedesign",
                "parent": "urn:subject:1",
              },
              Object {
                "id": "urn:topic:1_1",
                "name": "Mediedesign",
                "parent": "urn:topic:1",
              },
              Object {
                "contentUri": "urn:article:1_2",
                "id": "urn:topic:1_2",
                "name": "Idéutvikling",
                "parent": "urn:topic:1",
              },
              Object {
                "id": "urn:topic:1_2_1",
                "name": "Mediebransjen",
                "parent": "urn:topic:1_2",
              },
              Object {
                "id": "urn:topic:2",
                "name": "Mediekommunikasjon",
              },
            ],
          },
          "type": "SET_TOPICS",
        },
        "channel": null,
      },
    },
  ],
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "SELECT": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "SELECT": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "@@redux-saga-test-plan/json/function/actionCreator",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "@@redux-saga-test-plan/json/function/actionCreator",
      },
    },
  ],
}
`;

exports[`topicSagas watchFetchTopics when topics already are in state 1`] = `
Object {
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "SELECT": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "@@redux-saga-test-plan/json/function/actionCreator",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "@@redux-saga-test-plan/json/function/actionCreator",
      },
    },
  ],
}
`;

exports[`topicSagas watchFetchTopicsWithIntroductions when topics are in state 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "urn:subject:1",
          undefined,
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/fetchTopicsWithIntroductions",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {},
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Array [
            Object {
              "contentUri": "urn:article:1",
              "id": "urn:topic:1",
              "name": "Idéutvikling og mediedesign",
              "parent": "urn:subject:1",
            },
            Object {
              "id": "urn:topic:1_1",
              "name": "Mediedesign",
              "parent": "urn:topic:1",
            },
            Object {
              "contentUri": "urn:article:1_2",
              "id": "urn:topic:1_2",
              "name": "Idéutvikling",
              "parent": "urn:topic:1",
            },
            Object {
              "id": "urn:topic:1_2_1",
              "name": "Mediebransjen",
              "parent": "urn:topic:1_2",
            },
            Object {
              "id": "urn:topic:2",
              "name": "Mediekommunikasjon",
              "parent": undefined,
            },
          ],
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/fetchTopicIntroductions",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {},
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Array [
            "1",
            "1_2",
          ],
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": Object {
            "articleIntroductions": Array [
              Object {
                "id": "urn:article:1",
                "introduction": "Introduction 1",
              },
              Object {
                "id": "urn:article:1_2",
                "introduction": "Introduction 2",
              },
            ],
            "topics": Array [
              Object {
                "contentUri": "urn:article:1",
                "id": "urn:topic:1",
                "name": "Idéutvikling og mediedesign",
                "parent": "urn:subject:1",
              },
              Object {
                "id": "urn:topic:1_1",
                "name": "Mediedesign",
                "parent": "urn:topic:1",
              },
              Object {
                "contentUri": "urn:article:1_2",
                "id": "urn:topic:1_2",
                "name": "Idéutvikling",
                "parent": "urn:topic:1",
              },
              Object {
                "id": "urn:topic:1_2_1",
                "name": "Mediebransjen",
                "parent": "urn:topic:1_2",
              },
              Object {
                "id": "urn:topic:2",
                "name": "Mediekommunikasjon",
                "parent": undefined,
              },
            ],
          },
          "type": "SET_TOPIC_INTRODUCTIONS",
        },
        "channel": null,
      },
    },
  ],
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "SELECT": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "SELECT": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "@@redux-saga-test-plan/json/function/actionCreator",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "@@redux-saga-test-plan/json/function/actionCreator",
      },
    },
  ],
}
`;
